Sources/TensorFlow/Core/KeyPathIterable.swift

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exceptions.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//===----------------------------------------------------------------------===//
+// KeyPathIterable
+//===----------------------------------------------------------------------===//
+
+import _Differentiation
+
+#if TENSORFLOW_USE_STANDARD_TOOLCHAIN
+@_spi(Reflection) import Swift
+
+/// An implementation detail of `KeyPathIterable`; do not use this protocol
+/// directly.
+public protocol _KeyPathIterableBase {
+  var _allKeyPathsTypeErased: [AnyKeyPath] { get }
+  var _recursivelyAllKeyPathsTypeErased: [AnyKeyPath] { get }
+}
+
+/// A type whose values provides custom key paths to properties or elements.
+public protocol KeyPathIterable: _KeyPathIterableBase {
+  /// A type that can represent a collection of all key paths of this type.
+  associatedtype AllKeyPaths: Collection
+    where AllKeyPaths.Element == PartialKeyPath<Self>
+
+  /// A collection of all custom key paths of this value.
+  var allKeyPaths: AllKeyPaths { get }
+}
+
+public extension KeyPathIterable {
+  var allKeyPaths: [PartialKeyPath<Self>] {
+    var out = [PartialKeyPath<Self>]()
+    _forEachFieldWithKeyPath(of: Self.self, options: .ignoreUnknown) { name, kp in
+      out.append(kp)
+      return true
+    }
+    return out
+  }
```

Comment:
Note: sneaky default implementation - types still must declare a conformance to `KeyPathIterable` to use `KeyPathIterable.allKeyPaths`, unlike [this alternative implementation](https://github.com/apple/swift-evolution-staging/tree/reflection).

This seems reasonable though for maintaining the same API surface!

Reply:
Yes, I think that a standalone function is fine (ie: `var allKeyPaths: [PartialKeyPath<Self>]`) and people should just use that instead, but we still have code using KeyPathIterable.